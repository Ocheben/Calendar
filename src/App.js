import React from 'react';
import './App.scss';

const App = () => (
  <div className="App">
    <header className="App-header">Calendar</header>
  </div>
);
export default App;
