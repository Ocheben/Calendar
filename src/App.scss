  /* FONT IMPORTS */

  @import url(https://fonts.googleapis.com/css?family=Open+Sans:300,400,700);
  @import url(https://fonts.googleapis.com/icon?family=Material+Icons);

  .icon {
    font-family: 'Material Icons', serif;
    font-style: normal;
    display: inline-block;
    vertical-align: middle;
    line-height: 1;
    text-transform: none;
    letter-spacing: normal;
    word-wrap: normal;
    white-space: nowrap;
    direction: ltr;

    -webkit-font-smoothing: antialiased;
    text-rendering: optimizeLegibility;
    -moz-osx-font-smoothing: grayscale;
    font-feature-settings: 'liga';
  }


  /* VARIABLES */

  :root {
    --main-color: #1a8fff;
    --text-color: #777;
    --text-color-light: #ccc;
    --border-color: #eee;
    --bg-color: #f9f9f9;
    --neutral-color: #fff;
  }


  /* GENERAL */

  * {
    box-sizing: border-box;
    padding: 0;
    margin: 0;
  }

  body {
      font-family: 'Open Sans', 'Helvetica Neue', 'Helvetica', 'Arial', sans-serif;
      font-size: 1em;
      font-weight: 300;
      line-height: 1.5;
      color: var(--text-color);
      background: var(--bg-color);
      position: relative;
  }
  .App {
    background-color: #282c34;
    text-align: center;
  }
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
    height: 40vmin;
    pointer-events: none;
  }

  .root {
    height: 100vh;
  }

  header {
    display: block;
    width: 100%;
    padding: 1.75em 0.5em;
    border-bottom: 1px solid var(--border-color);
    background: var(--neutral-color);
  }

  header #logo {
    font-size: 175%;
    text-align: center;
    color: var(--main-color);
    line-height: 1;
  }

  header #logo .icon {
    padding-right: .25em;
  }

  main {
    display: block;
    margin: 0 auto;
    margin-top: 5em;
    max-width: 50em;
  }


  /* GRID */

  .row {
    margin: 0;
    padding: 0;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    width: 100%;
  }

  .row-middle {
    align-items: center;
  }

  .col {
    flex-grow: 1;
    flex-basis: 0;
    max-width: 100%;
  }

  .col-start {
    justify-content: flex-start;
    text-align: left;
  }

  .col-center {
    justify-content: center;
    text-align: center;
  }

  .col-end {
    justify-content: flex-end;
    text-align: right;
  }


  /* Calendar */

  .calendar {
    display: block;
    position: relative;
    width: 100%;
    background: var(--neutral-color);
    border: 1px solid var(--border-color);
  }

  .calendar .header {
    text-transform: uppercase;
    height: 12vh;
    font-weight: 700;
    font-size: 115%;
    padding: 1.5em 0.5em;
    border-bottom: 1px solid var(--border-color);
  }

  .calendar .header .icon {
    cursor: pointer;
    transition: .15s ease-out;
    &:hover {
      transform: scale(1.75);
      transition: .25s ease-out;
      color: var(--main-color);
    }
  }

  .calendar .header .icon:first-of-type {
    margin-left: 1em;
  }

  .calendar .header .icon:last-of-type {
    margin-right: 1em;
  }

  .calendar .days {
    align-content: center;
    height: 4.7vh;
    text-transform: uppercase;
    font-weight: 400;
    color: var(--text-color-light);
    font-size: 70%;
    padding: .75em 0;
    border-bottom: 1px solid var(--border-color);
  }

  .calendar .body .cell {
    position: relative;
    height: 16.7vh;
    border-right: 1px solid var(--border-color);
    overflow: hidden;
    cursor: auto;
    background: var(--neutral-color);
    transition: 0.25s ease-out;
    &:hover {
      background: var(--bg-color);
    transition: 0.5s ease-out;
    }
  }

  .calendar .body .selected {
    border-left: 10px solid transparent;
    border-image: linear-gradient(45deg, #1a8fff 0%,#53cbf1 40%);
    border-image-slice: 1;
  }

  .calendar .body .row {
    border-bottom: 1px solid var(--border-color);
  }

  .calendar .body .row:last-child {
    border-bottom: none;
  }

  .calendar .body .cell:last-child {
    border-right: none;
  }

  .calendar .body .cell .number {
    position: relative;
    font-size: 82.5%;
    line-height: 1;
    font-weight: 700;
  }

  .calendar .body .disabled {
    color: var(--text-color-light);
    pointer-events: none;
  }

  .calendar .body .cell .bg {
    font-weight: 700;
    line-height: 1;
    color: var(--main-color);
    opacity: 0;
    font-size: 8em;
    position: absolute;
    top: -.2em;
    right: -.05em;
    transition: .25s ease-out;
    letter-spacing: -.07em;
  }

  .calendar .body .cell:hover .bg, .calendar .body .selected .bg  {
    opacity: 0.05;
    transition: .5s ease-in;
  }

  .calendar .body .col {
    flex-grow: 0;
    flex-basis: calc(100%/7);
    width: calc(100%/7);
  }

  .modal {
    display: block;
    position: absolute;
    top: 50%;
    left: 50%;
    background: #ffffff;
    box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
    width: 40%;
    height: 40%;
    transform: translate(-50%, -50%);
    border-radius: 0.3rem;
    animation-name: animatetop;
    animation-duration: 1s
  }
  @keyframes animatetop {
    from {opacity: 0}
    to {opacity: 1}
  }

  .modal-header {
    align-content: center;
    height: 15%;
    border-bottom: 1px solid #e9e9e9;
    bottom: 0;
    padding: 0.75em;
  }

  .modal-header h3 {
    display: inline;
  }

  .modal-header .icon {
    float: right;
    cursor: pointer;
    &:hover{
      transform: scale(1.75);
      transition: .25s ease-out;
      color: var(--main-color);
    }
  }

  .modal-content {
    padding: 1em;
    justify-content: center;
  }


  /* Input Style */
  .group {
    position:relative;
    margin: 0.7em auto;
    width: 70%;
    align-self: center;
  }

  select {
    background: #ffffff;
    margin: 0.4em;
    border: none;
    width: 100%;
    border-bottom:1px solid #757575;
    font-size:15px;
  }
  .select {
    width: 70%;
    margin: 1.4em auto;
  }
  input	{
    font-size:15px;
    padding:10px 10px 10px 5px;
    display:block;
    width:100%;
    border:none;
    border-bottom:1px solid #757575;
  }
  input:focus, select:focus	{ outline:none; }

  /* LABEL ======================================= */
  label 				 {
    color:#999;
    font-size:15px;
    font-weight:normal;
    position:absolute;
    pointer-events:none;
    left:5px;
    top:10px;
    transition:0.2s ease all;
    -moz-transition:0.2s ease all;
    -webkit-transition:0.2s ease all;
  }

  /* active state */
  input:focus ~ label, input:valid ~ label, select:focus ~ label, select:valid ~ label 		{
    top:-12px;
    font-size:13px;
    color:#5264AE;
  }

  /* BOTTOM BARS ================================= */
  .bar 	{ position:relative; display:block; width:100%; }
  .bar:before, .bar:after 	{
    content:'';
    height:2px;
    width:0;
    bottom:1px;
    position:absolute;
    background:#5264AE;
    transition:0.2s ease all;
    -moz-transition:0.2s ease all;
    -webkit-transition:0.2s ease all;
  }
  .bar:before {
    left:50%;
  }
  .bar:after {
    right:50%;
  }

  /* active state */
  input:focus ~ .bar:before, input:focus ~ .bar:after, select:focus ~ .bar:before, select:focus ~ .bar:after {
    width:50%;
  }

  /* HIGHLIGHTER ================================== */
  .highlight {
    position:absolute;
    height:60%;
    width:100px;
    top:25%;
    left:0;
    pointer-events:none;
    opacity:0.5;
  }

  /* active state */
  input:focus ~ .highlight, select:focus ~ .highlight {
    -webkit-animation:inputHighlighter 0.3s ease;
    -moz-animation:inputHighlighter 0.3s ease;
    animation:inputHighlighter 0.3s ease;
  }

  /* ANIMATIONS ================ */
  @-webkit-keyframes inputHighlighter {
    from { background:#5264AE; }
    to 	{ width:0; background:transparent; }
  }
  @-moz-keyframes inputHighlighter {
    from { background:#5264AE; }
    to 	{ width:0; background:transparent; }
  }
  @keyframes inputHighlighter {
    from { background:#5264AE; }
    to 	{ width:0; background:transparent; }
  }

  .reminder {
    border-radius: 0.4em;
    margin: 0.3em 0.6em;
    cursor: pointer;
  }

  .reminder span {
    color: #ffffff;
    cursor: pointer;
  }

  .actions {
    // position: absolute;
   margin: 2em auto;
  }

  .actions button {
    border-radius: 0.3em;
    height: 3em;
    color: #ffffff;
    cursor: pointer;
  }

  .add-btn {
    background: #00897b;
    width: 50%;
  }

  .edit-btn {
    background: #1e88e5;
    width: 30%;
  }

  .del-btn {
    background: #e53935;
    width: 30%;
  }


