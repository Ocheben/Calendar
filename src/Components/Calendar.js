/* eslint-disable no-plusplus */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */
import React, { useState } from 'react';
import dateFns from 'date-fns';

const Calendar = () => {
  const [state, setState] = useState({
    currentMonth: new Date(),
    selectedDate: new Date()
  });

  const renderHeader = () => {
    const dateFormat = 'MMMM YYYY';

    return (
      <div className="header row flex-middle">
        <div className="col col-start">
          <div className="icon" onClick={prevMonth}>
            chevron_left
          </div>
        </div>
        <div className="col col-center">
          <span>{dateFns.format(state.currentMonth, dateFormat)}</span>
        </div>
        <div className="col col-end" onClick={nextMonth}>
          <div className="icon">chevron_right</div>
        </div>
      </div>
    );
  };

  const renderDays = () => {
    const dateFormat = 'dddd';
    const days = [];

    const startDate = dateFns.startOfWeek(state.currentMonth);

    for (let i = 0; i < 7; i++) {
      days.push(
        <div className="col col-center" key={i}>
          {dateFns.format(dateFns.addDays(startDate, i), dateFormat)}
        </div>
      );
    }

    return <div className="days row">{days}</div>;
  };

  const renderCells = () => {
    const { currentMonth, selectedDate } = state;
    const monthStart = dateFns.startOfMonth(currentMonth);
    const monthEnd = dateFns.endOfMonth(monthStart);
    const startDate = dateFns.startOfWeek(monthStart);
    const endDate = dateFns.endOfWeek(monthEnd);

    const dateFormat = 'D';
    const rows = [];

    let days = [];
    let day = startDate;
    let formattedDate = '';

    while (day <= endDate) {
      for (let i = 0; i < 7; i++) {
        formattedDate = dateFns.format(day, dateFormat);
        const cloneDay = day;
        days.push(
          <div
            className={`col cell ${
              // eslint-disable-next-line no-nested-ternary
              !dateFns.isSameMonth(day, monthStart)
                ? 'disabled'
                : dateFns.isSameDay(day, selectedDate) ? 'selected' : ''
            }`}
            key={day}
            onClick={() => onDateClick(dateFns.parse(cloneDay))}
          >
            <span className="number">{formattedDate}</span>
            <span className="bg">{formattedDate}</span>
          </div>
        );
        day = dateFns.addDays(day, 1);
      }
      rows.push(
        <div className="row" key={day}>
          {days}
        </div>
      );
      days = [];
    }
    return <div className="body">{rows}</div>;
  };

  const onDateClick = (day) => {
    setState({
      ...state,
      selectedDate: day
    });
  };

  const nextMonth = () => {
    setState({
      ...state,
      currentMonth: dateFns.addMonths(state.currentMonth, 1)
    });
  };

  const prevMonth = () => {
    setState({
      ...state,
      currentMonth: dateFns.subMonths(state.currentMonth, 1)
    });
  };

  return (
    <div className="calendar">
      {renderHeader()}
      {renderDays()}
      {renderCells()}
    </div>
  );
};

export default Calendar;
