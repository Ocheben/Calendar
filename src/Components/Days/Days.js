/* eslint-disable no-plusplus */
import React from 'react';
import * as dateFns from 'date-fns';

const Days = (props) => {
  const { currentMonth } = props;
  const dateFormat = 'EEEE';
  const days = [];

  const startDate = dateFns.startOfWeek(currentMonth);

  for (let i = 0; i < 7; i++) {
    days.push(
      <div className="col col-center" key={i}>
        {dateFns.format(dateFns.addDays(startDate, i), dateFormat)}
      </div>
    );
  }

  return <div className="days row">{days}</div>;
};

export default Days;
