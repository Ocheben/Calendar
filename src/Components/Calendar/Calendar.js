import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import DateFnsUtils from '@date-io/date-fns';
import * as dateFns from 'date-fns';
import apiCall from '../../_services/requests';
import { storeReminders } from '../../_store/actions';
import Cells from '../Cells/Cells';
import Header from '../Header/Header';
import Days from '../Days/Days';
import Modal from '../Modal/Modal';

const Calendar = (props) => {
  const { reminders, dispatch } = props;
  const [state, setState] = useState({
    currentMonth: new Date(),
    selectedDate: new Date(),
    modalOpen: false,
    formInputs: {
      time: new Date()
    },
    isEdit: false
  });

  const refreshReminders = async () => {
    const response = await apiCall('GET', 'http://localhost:3001/reminders');
    if (response.constructor === Array) {
      dispatch(storeReminders(response));
    }
  };

  useEffect(() => {
    const getReminders = async () => {
      const response = await apiCall('GET', 'http://localhost:3001/reminders');
      if (response.constructor === Array) {
        dispatch(storeReminders(response));
      }
    };
    getReminders();
  }, [dispatch]);

  const nextMonth = () => {
    setState({
      ...state,
      currentMonth: dateFns.addMonths(state.currentMonth, 1)
    });
  };

  const prevMonth = () => {
    setState({
      ...state,
      currentMonth: dateFns.subMonths(state.currentMonth, 1)
    });
  };

  const openModal = (modalState, formInputs, isEdit) => {
    setState({
      ...state,
      modalOpen: modalState,
      formInputs,
      isEdit
    });
  };

  const handleDateChange = (date) => {
    setState({
      ...state,
      formInputs: {
        ...state.formInputs,
        time: date
      }
    });
  };
  const handleChange = ({ target }) => {
    const { name, value } = target;
    setState({
      ...state,
      formInputs: {
        ...state.formInputs,
        [name]: value
      }
    });
  };

  const handleSubmit = async () => {
    await apiCall('POST', 'http://localhost:3001/reminders', state.formInputs);
    refreshReminders();
    openModal(false, {}, false);
  };

  const handleEdit = async () => {
    await apiCall('PUT', `http://localhost:3001/reminders/${state.formInputs.id}`, state.formInputs);
    refreshReminders();
    openModal(false, {}, false);
  };

  const handleDelete = async () => {
    await apiCall('DELETE', `http://localhost:3001/reminders/${state.formInputs.id}`);
    refreshReminders();
    openModal(false, {}, false);
  };

  const { formInputs } = state;
  return (
    <MuiPickersUtilsProvider utils={DateFnsUtils}>
      <div className="root">
        <div className="calendar">
          <Header
            prevMonth={prevMonth}
            nextMonth={nextMonth}
            openModal={openModal}
            currentMonth={state.currentMonth}
          />
          <Days currentMonth={state.currentMonth} />
          <Cells
            currentMonth={state.currentMonth}
            selectedDate={state.selectedDate}
            reminders={reminders}
            openModal={openModal}
          />
        </div>
        {
        state.modalOpen
        && (
        <Modal
          formInputs={formInputs}
          isEdit={state.isEdit}
          openModal={openModal}
          handleChange={handleChange}
          handleDateChange={handleDateChange}
          handleSubmit={handleSubmit}
          handleEdit={handleEdit}
          handleDelete={handleDelete}
        />
        )
      }

      </div>
    </MuiPickersUtilsProvider>
  );
};

const mapStateToProps = state => ({
  reminders: state.reminders
});

export default connect(mapStateToProps)(Calendar);
