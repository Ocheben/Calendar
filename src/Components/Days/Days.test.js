import React from 'react';
import { shallow } from 'enzyme';
import Days from './Days';
import Calendar from '../Calendar/Calendar';
import App from '../../App';

describe('Days', () => {
  it('should render correctly', () => {
    const component = shallow(
      <App>
        <Calendar><Days /></Calendar>
      </App>
    );

    expect(component).toMatchSnapshot();
  });
});
