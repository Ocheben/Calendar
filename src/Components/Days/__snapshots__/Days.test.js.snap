// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Days should render correctly 1`] = `
<div
  className="App"
>
  <Connect(Calendar) />
</div>
`;
